### ISSUE TITLE
Correctness?

### ISSUE THREAD

**Issue by isarandi** [Fdevmsy/PyTorch-Soft-Argmax#2]
Based on just reading the code, I suspect the implementation is not correct. It seems to first unroll the volume into a single dimension, then applies one-dimensional soft-argmax and unravels the resulting single index to three indices corresponding to the volume's axes., However, this is not how soft-argmax should operate in multiple dimensions.
For each coordinate (x, y, z), you first need to sum the volume along the other spatial axes, then apply one-dimensional soft-argmax to the resulting array to obtain your coordinate along that axis.

Something like this:
```
        arange = torch.arange(0, cube_side)
        net = F.softmax(net.reshape(*net.shape[:-3], -1), dim=-1)
        net = net.reshape(*net.shape[:-3], cube_side, cube_side, cube_side)
        i = torch.sum(torch.sum(net, dim=(-2, -1)) * arange, dim=-1, keepdim=True)
        j = torch.sum(torch.sum(net, dim=(-3, -1)) * arange, dim=-1, keepdim=True)
        k = torch.sum(torch.sum(net, dim=(-3, -2)) * arange, dim=-1, keepdim=True)
        return torch.cat([vec_x, vec_y, vec_z], dim=-1)
```

**Comment by qiufeng1994** [comment]
Impressive! Thanks!
